pkg/agent/controller/networkpolicy/multicast.go

```diff
@@ -0,0 +1,337 @@
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleAdd,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) updateGroupAddressForTableIDs(ruleID string, priority *uint16, mcastGroupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		priority:     priority,
+		groupAddress: mcastGroupAddresses,
+		event:        ruleUpdate,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) deleteGroupAddressForTableIDs(ruleID string, groupAddresses []string) {
+	igmpRule := &mcastIGMPRule{
+		ruleID:       ruleID,
+		groupAddress: groupAddresses,
+		event:        ruleDelete,
+	}
+	c.queue.Add(igmpRule)
+}
+
+func (c *multicastController) validate(iface *interfacestore.InterfaceConfig,
+	groupAddress net.IP, direction crdv1beta.Direction) (*v1alpha1.RuleAction, apitypes.UID, *crdv1beta.NetworkPolicyType, string, error) {
+
+	var ruleTypePtr *crdv1beta.NetworkPolicyType
+	action, uuid, ruleName := v1alpha1.RuleActionAllow, apitypes.UID(""), ""
+	if iface == nil {
+		// Check the if iface exists
+		klog.ErrorS(fmt.Errorf("iface should not be empty"), "")
+		return nil, "", nil, "", fmt.Errorf("iface should not be empty")
+	}
+	ns, podname := iface.PodNamespace, iface.PodName
+	obj, exists, _ := c.mcastItemRuleCache.GetByKey(groupAddress.String())
+	if !exists {
+		klog.V(2).InfoS("Rule for group does not exist", "group", groupAddress.String())
+		action = v1alpha1.RuleActionAllow
+		return &action, "", nil, "", nil
+	}
+	item := obj.(*mcastItem)
+	var matchedRule *CompletedRule
+	member := &crdv1beta.GroupMember{
+		Pod: &crdv1beta.PodReference{
+			Name:      podname,
+			Namespace: ns,
+		},
+	}
+	for ruleID := range item.ruleIDs {
+		// Iterate all rules relate to this multicast group address
+		r, _, _ := c.ruleCache.GetCompletedRule(ruleID)
+		// Find the rule with highest priority
+		if direction == r.Direction && r.TargetMembers.Has(member) == true {
+			if (matchedRule == nil) || *(item.ruleIDs[ruleID]) > *(item.ruleIDs[matchedRule.ID]) {
+				matchedRule = r
+			}
+		}
+	}
+	if matchedRule != nil {
+		ruleTypePtr = new(crdv1beta.NetworkPolicyType)
+		action, uuid, *ruleTypePtr, ruleName = *matchedRule.Action, matchedRule.PolicyUID, matchedRule.SourceRef.Type, matchedRule.Name
+	} else {
+		action, uuid, ruleName = v1alpha1.RuleActionAllow, "", ""
+	}
+	klog.V(2).InfoS("Call validation:", "action", action, "uuid", uuid, "ruleName", ruleName, "ruleType", ruleTypePtr)
+	return &action, uuid, ruleTypePtr, ruleName, nil
+}
+
+func (c *multicastController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *multicastController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	defer c.queue.Done(obj)
+
+	// We expect igmpRule to come off the workqueue.
+	if item, ok := obj.(*mcastIGMPRule); !ok {
```

Comment:
This is now how workqueue is supposed to be used. What you need is just a channel, and perhaps you don't even need a channel and could just process update immediately if it just updates some cache.

Reply:
changed to channel